### PR TITLE
support permission (bits of permission and excution) and support  symlinks

### PR DESCRIPTION
**Pull request by Lambourl**
The purpose of this pr is to have the permission and execution bits and to restore these permissions on extraction.
Also we can now use symlinks and store them.

### CODE REVIEW

File: src/repo/pack.rs [hunk 1]
```diff
@@ -492,15 +503,51 @@ fn verify_object(buf: &[u8], exp_checksum: &ObjectChecksum) -> Result<(), Error>
...
+    if symlink_target.is_file() {
+        symlink(symlink_target, path)?;
+    } else {
+        let mut _f = create_file(symlink_target)?;
+        symlink(symlink_target, path)?;
+        fs::remove_file(symlink_target)?;
+    }
+    Ok(())
+
+}
```
Comment (pysco68) [inline src/repo/pack.rs]: 1. Function name is *meh* (safety != safely)
2. effective difference between both is now the distinction of dir symlinks vs file symlinks. Single fn => less oportunity for fuckups
3. do we know if the target shall be a dir/file? If yes, then the create "safely" method should make that distinction on windows
4. if symlinking fails we have a leftover placeholder file => RAII / try-catch

File: src/repo/pack.rs [hunk 2]
```diff
@@ -492,15 +503,51 @@ fn verify_object(buf: &[u8], exp_checksum: &ObjectChecksum) -> Result<(), Error>
...
+
+
 /// Writes the object to the specified path, taking care
 /// of adjusting file permissions.
-fn write_object(buf: &[u8], path: &Path, last_modified:i64, last_modified_nanos:u32 ) -> Result<(), Error> {
+fn write_object(buf: &[u8], path: &Path, last_modified:i64, last_modified_nanos:u32, bits_mods:u32, is_symlink:bool, symlink_target:PathBuf ) -> Result<(), Error> {
```
Comment (pysco68) [inline src/repo/pack.rs]: i know that you just basically extended the existing params list, but why not pass the entries' `metadata` struct?

File: src/repo/pack.rs [hunk 3]
```diff
@@ -668,7 +718,7 @@ fn extract_files(
             path_buf.clear();
             path_buf.push(&output_dir);
             path_buf.push(&entry.path);
-            stats.write_time += measure_ok(|| write_object(&buf[..], &path_buf, entry.metadata.last_modified, entry.metadata.last_modified_nanos))?
+            stats.write_time += measure_ok(|| write_object(&buf[..], &path_buf, entry.metadata.last_modified, entry.metadata.last_modified_nanos,entry.metadata.bits_mods,entry.metadata.is_symlink_file,entry.metadata.symlink_target))?
```
Comment (pysco68) [inline src/repo/pack.rs]: See previous comment

File: src/repo/repository.rs [hunk 4]
```diff
@@ -531,22 +538,47 @@ impl Repository {
...
+
             let buf = fs::read(&file_path)?;
             let mut checksum = [0u8; 20];
             let mut hasher = Sha1::new();
             hasher.input(&buf);
             hasher.result(&mut checksum);
             self.write_loose_object(&*buf, &temp_dir, &checksum)?;
 
+            //let path = Path::new(&file_path);
+            let is_symlink_files = Path::new(&file_path).is_symlink();
```
Comment (pysco68) [inline src/repo/repository.rs]: plural?

File: src/repo/repository.rs [hunk 5]
```diff
@@ -531,22 +538,47 @@ impl Repository {
...
+                    last_modified: FileTime::from_last_modification_time(&metadata).seconds()-11644473600,
+                    #[cfg(target_family = "unix")]
                     last_modified: FileTime::from_last_modification_time(&metadata).seconds(),
                     last_modified_nanos: FileTime::from_last_modification_time(&metadata).nanoseconds(),
+                    #[cfg(target_family = "unix")]
+                    bits_mods: metadata.permissions().mode(),
+                    #[cfg(target_family = "windows")]
+                    bits_mods: 0o777,
+                    is_symlink_file: is_symlink_files,
+                    symlink_target: symlink_target,
```
Comment (pysco68) [inline src/repo/repository.rs:583]: We should probably keep track of the type of the symlink target (dir/file) for correct dummy creation on windows

File: src/repo/repository.rs [hunk 6]
```diff
@@ -570,6 +602,25 @@ impl Repository {
...
+        for b in file_path_replaced.bytes() {
+            if b == 13 {
+                is_cr = true;
+            }
+        }
+        let mut file_path_replacedop: &str = &file_path_replaced;
+
+        if is_cr == true {
+            file_path_replacedop = &file_path_replaced[0..file_path_replaced.len() - 1];
+        }
```
Comment (pysco68) [inline src/repo/repository.rs]: is that kind of a badly implemented string.trim_end('\r')?

Comment (pysco68) [reply src/repo/repository.rs]: I mean it does look for \r in the whole string and then proceeds to replace the last char.... even if it's not a \r 🤷

File: src/repo/repository.rs [hunk 7]
```diff
@@ -968,10 +1062,19 @@ fn is_elfshaker_data_path(p: &Path) -> bool {
...
-    };
+    pub fn get_example_md () -> ObjectMetadata{
+        let example_md: ObjectMetadata = ObjectMetadata {
+            size: 1,
+            offset: LOOSE_OBJECT_OFFSET,
+            last_modified: 0,
+            last_modified_nanos: 0,
+            bits_mods: 0,
+            is_symlink_file: false,
+            symlink_target:  Path::new("/test/haha/foo.txt").to_path_buf(),
```
Comment (pysco68) [inline src/repo/repository.rs]: `is_symlink_file == false` and a set `symlink_target` ?!

**Review by pysco68 — CHANGES_REQUESTED** [review]
The bug you thought fixed twice now is a bit more complex than what I think you understood.

Basically Elfshaker is parallelizing the pack-/unpacking over multiple threads, so really the issue is not one of "order" but parallelity (and order per bucket of entries to extract/compress in a given snapshot).

See `src/repo/pack.rs:462` 

The "safe" symlink creation is half broken too, sadly, as it only accounts for file vs dir-type symlinks on windows in the case of the "target already exists" case (the symlink type is determined on creation and does not automatically change by the target type because it's a NTFS remount-point that is pointing either to a node (dir) OR a leaf (file) and the handling seems to be different internally.

File: src/repo/repository.rs (hunk 5 above)
Comment (Lambourl) [reply src/repo/repository.rs:583]: We had the discussion this morning we can't get a dir

File: src/repo/pack.rs (hunk 1 above)
Comment (Lambourl) [reply src/repo/pack.rs]: done

File: src/repo/pack.rs (hunk 2 above)
Comment (Lambourl) [reply src/repo/pack.rs]: done

File: src/repo/pack.rs (hunk 3 above)
Comment (Lambourl) [reply src/repo/pack.rs]: done

File: src/repo/repository.rs (hunk 4 above)
Comment (Lambourl) [reply src/repo/repository.rs]: done

File: src/repo/repository.rs (hunk 6 above)
Comment (Lambourl) [reply src/repo/repository.rs]: done

File: src/repo/repository.rs (hunk 7 above)
Comment (Lambourl) [reply src/repo/repository.rs]: done with a empty pathbuf

**Pull request closed without merging**